### PR TITLE
Add Challenges

### PR DESCRIPTION
**Pull request by Xavion3**
Implements basic challenges. This is still incomplete, but this is being created in draft mode for now so people can see it.

### CODE REVIEW

File: src/data/challenge.ts [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+import Pokemon from "#app/field/pokemon.js";
+import { BattleType, FixedBattleConfig } from "#app/battle.js";
+import { TrainerType } from "./enums/trainer-type";
+import Trainer, { TrainerVariant } from "#app/field/trainer.js";
+import { GameMode } from "#app/game-mode.js";
+
+export enum ChallengeType {
+   /** Challenges which modify what starters you can choose
+    * @param args [0] {@link PokemonSpecies} The species to check
+    *             [1] {@link Utils.BooleanHolder} Sets to false if illegal, pass in true.
```
Comment (Tempo-anon) [inline src/data/challenge.ts:15]: Enums can have params? I thought by default they were just 0, 1, 2, and so on

File: src/data/challenge.ts [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+    throw new Error("Method not implemented! Use derived class");
+  }
+}
+
+type ChallengeCondition = (data: GameData) => boolean;
+
+/**
+ * Implements a mono generation challenge.
+ */
+export class SingleGenerationChallenge extends Challenge {
```
Comment (Tempo-anon) [inline src/data/challenge.ts:246]: nit: Should all the `Challenge`s go in their own file instead of `challenge.ts`?

File: src/data/challenge.ts [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+      if (species.generation !== this.value) {
+        isValidStarter.value = false;
+        return true;
+      }
+      break;
+    case ChallengeType.POKEMON_IN_BATTLE:
+      const pokemon = args[0] as Pokemon;
+      const isValidPokemon = args[1] as Utils.BooleanHolder;
+      if (pokemon.isPlayer() && ((pokemon.species.formIndex === 0 ? pokemon.species : getPokemonSpecies(pokemon.species.speciesId)).generation !== this.value || (pokemon.isFusion() && (pokemon.fusionSpecies.formIndex === 0 ? pokemon.fusionSpecies : getPokemonSpecies(pokemon.fusionSpecies.speciesId)).generation !== this.value))) {
+        isValidPokemon.value = false;
```
Comment (Tempo-anon) [inline src/data/challenge.ts:272]: Please add a comment here describing the valid logic

File: src/data/challenge.ts [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+  static loadChallenge(source: SingleTypeChallenge | any): SingleTypeChallenge {
+    const newChallenge = new SingleTypeChallenge();
+    newChallenge.value = source.value;
+    newChallenge.severity = source.severity;
+    return newChallenge;
+  }
+}
+
+/**
+ * Implements a mono type challenge.
```
Comment (Tempo-anon) [inline src/data/challenge.ts:374]: Wrong comment?

File: src/locales/de/challenges.ts
Comment (Tempo-anon) [inline src/locales/de/challenges.ts:1]: Localization is good but really do bloat up PRs. Maybe something to discuss with the other devs on if localization for features should be in their own PR

File: src/locales/en/achv.ts [hunk 5]
```diff
@@ -168,4 +168,99 @@ export const achv: AchievementTranslationEntries = {
     name: "Undefeated",
     description: "Beat the game in classic mode",
   },
+
+  "MONO_GEN_ONE": {
+    name: "The Original Rival",
```
Comment (Tempo-anon) [inline src/locales/en/achv.ts:173]: I think the names should reference the gen/region rather than the final champion. For this one I'd either go with something like "Origins" for a more serious note and "Smell ya later" if going for more levity

File: src/locales/en/challenges.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import { SimpleTranslationEntries } from "#app/plugins/i18n";
+
+export const challenges: SimpleTranslationEntries = {
+  "title": "Challenge Modifiers",
+  "points": "Bad Ideas",
```
Comment (Tempo-anon) [inline src/locales/en/challenges.ts:5]: Maybe `Challenge Points` to make it clearer for players?

File: src/phases.ts [hunk 7]
```diff
@@ -1287,20 +1310,20 @@ export class SummonPhase extends PartyMemberPokemonPhase {
...
+    if (!partyMember.isAllowedInBattle()) {
+      console.warn("The Pokemon about to be sent out is fainted or illegal under a challenge. Attempting to resolve...");
       const party = this.getParty();
 
       // Find the first non-fainted Pokemon index above the current one
-      const nonFaintedIndex = party.findIndex((p, i) => i > this.partyMemberIndex && !p.isFainted());
-      if (nonFaintedIndex === -1) {
+      const legalIndex = party.findIndex((p, i) => i > this.partyMemberIndex && p.isAllowedInBattle());
+      if (legalIndex === -1) {
         console.error("Party Details:\n", party);
```
Comment (Tempo-anon) [inline src/phases.ts:1321]: This just freezes and doesn't send the player to a game over?

File: src/system/game-data.ts [hunk 8]
```diff
@@ -124,6 +125,7 @@ export interface SessionSaveData {
   trainer: TrainerData;
   gameVersion: string;
   timestamp: integer;
+  challenges: ChallengeData[];
```
Comment (Tempo-anon) [inline src/system/game-data.ts:128]: Will there need to be a change to the server so this data is actually saved and loaded?

File: src/ui/challenges-select-ui-handler.ts [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+  }
+
+  setup() {
+    const ui = this.getUi();
+
+    this.challengesContainer = this.scene.add.container(1, -(this.scene.game.canvas.height / 6) + 1);
+
+    this.challengesContainer.setInteractive(new Phaser.Geom.Rectangle(0, 0, this.scene.game.canvas.width / 6, this.scene.game.canvas.height / 6), Phaser.Geom.Rectangle.Contains);
+
+    // TODO: Change this back to /9 when adding in difficulty
```
Comment (Tempo-anon) [inline src/ui/challenges-select-ui-handler.ts:42]: I assume all the commented out stuff is related to difficulty which is not implemented yet?

Comment (Xavion3) [reply src/ui/challenges-select-ui-handler.ts:42]: Correct, the difficulty code is related to the full challenges system. For now this is just trying to get the basics in and make monotype and monogen playable.

File: src/data/challenge.ts (hunk 2 above)
Comment (Xavion3) [reply src/data/challenge.ts:246]: I'd say no unless we also want to restructure things like moves and abilities to do the same as well.

File: src/data/challenge.ts (hunk 1 above)
Comment (Xavion3) [reply src/data/challenge.ts:15]: They can't, so the info here is to describe the parameters when using the apply function for the various challenge types.

File: src/data/challenge.ts (hunk 2 above)
Comment (Tempo-anon) [reply src/data/challenge.ts:246]: Honestly I wouldn't be opposed to doing that, but I don't know if there's a clear cut way to breaking up the giant abilities or moves file. Here though, each challenge should be distinct enough

File: src/phases.ts (hunk 7 above)
Comment (Xavion3) [reply src/phases.ts:1321]: I mean yes, but that's because the game would be in an error state. It's the same error that happens if you try to switch to a fainted pokemon we already have. Do we want to delete peoples saves when that happens?

File: src/locales/en/achv.ts (hunk 5 above)
Comment (Xavion3) [reply src/locales/en/achv.ts:173]: I don't mind some champion references, but I'd need alternative ideas for names. I've considered something like `151` for gen one.

File: src/locales/en/challenges.ts (hunk 6 above)
Comment (Xavion3) [reply src/locales/en/challenges.ts:5]: Oh yeah, this was an early joke-y name.

File: src/data/challenge.ts (hunk 3 above)
Comment (Xavion3) [reply src/data/challenge.ts:272]: It's the mono generation challenge, it checks if the pokemon is from the correct generation. I'm not sure what you could assume `.generation !== this.value` to mean other than that.

File: src/data/challenge.ts (hunk 4 above)
Comment (Xavion3) [reply src/data/challenge.ts]: Fixed

File: src/data/challenge.ts (hunk 2 above)
Comment (Xavion3) [reply src/data/challenge.ts:255]: I feel like that'll just massively bloat file count to give them their own challenges too, remember while there are only really two here there's plans for dozens.

File: src/system/game-data.ts (hunk 8 above)
Comment (XGSleepWalker) [reply src/system/game-data.ts:128]: https://github.com/pagefaultgames/rogueserver/pull/34

**Review by Tempo-anon — APPROVED** [review]
Fill out the PR description and title please

**Review by patapancakes — APPROVED** [review]

**Comment by Simplyfenton** [comment]
Would it be possible to merge the functionality of Transfer Items and Check Team? Currently the changes made in [this commit](https://github.com/pagefaultgames/pokerogue/pull/1459/commits/d9a65eff1356da70c4ed303dfed9d99581c9056b#diff-bdf7f0eba287c240a7aa81f9dbd07af7aa8579cb374585985f3e1c9c43b7fd74) conflict heavily with [my PR](https://github.com/pagefaultgames/pokerogue/pull/1093/) in such a way that that's likely the only solution that would make my PR viable without reverting back to the less ideal UI handling around the lower shop options. I also don't much want to make any changes to my PR without asking first, since the work needed to resolve the conflicts could be pointless if my PR isn't valuable/being considered in the first place.
